File: backend/app/api/attachments.py
from collections.abc import Iterator
from urllib.parse import quote

from fastapi import APIRouter, Depends, File, Form, UploadFile, status
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

from ..authorization import Principal
from ..config import get_settings
from ..database import get_db
from ..malware_scanner import get_malware_scanner
from ..repositories.attachments import AttachmentRepository
from ..repositories.identity import AuditRepository
from ..repositories.postgres import PostgreSQLTicketRepository
from ..schemas import AttachmentDelete, AttachmentList, AttachmentRead, SuccessResponse
from ..services.attachment_service import AttachmentService
from ..storage import get_object_storage
from .dependencies import get_user_principal


router = APIRouter(tags=["attachments"])


def get_attachment_service(db: Session = Depends(get_db)) -> AttachmentService:
    return AttachmentService(
        PostgreSQLTicketRepository(db),
        AttachmentRepository(db),
        AuditRepository(db),
        get_object_storage(),
        get_malware_scanner(),
        get_settings(),
    )


@router.post(
    "/api/v1/tickets/{ticket_id}/attachments",
    response_model=SuccessResponse[AttachmentRead],
    status_code=status.HTTP_201_CREATED,
)
def upload_attachment(
    ticket_id: str,
    file: UploadFile = File(...),
    attachment_type: str = Form(...),
    visibility: str = Form("public"),
    principal: Principal = Depends(get_user_principal),
    service: AttachmentService = Depends(get_attachment_service),
):
    result = service.upload(
        ticket_id, principal, file.file, file.filename, file.content_type,
        attachment_type.strip(), visibility.strip(),
    )
    return SuccessResponse(data=result)


@router.get(
    "/api/v1/tickets/{ticket_id}/attachments",
    response_model=SuccessResponse[AttachmentList],
)
def list_attachments(
    ticket_id: str,
    principal: Principal = Depends(get_user_principal),
    service: AttachmentService = Depends(get_attachment_service),
):
    return SuccessResponse(data=service.list(ticket_id, principal))


def _stream_reader(reader) -> Iterator[bytes]:
    try:
        while True:
            chunk = reader.read(1024 * 1024)
            if not chunk:
                break
            yield chunk
    finally:
        close = getattr(reader, "close", None)
        if close:
            close()
        release = getattr(reader, "release_conn", None)
        if release:
            release()


@router.get("/api/v1/attachments/{attachment_id}/download")
def download_attachment(
    attachment_id: str,
    principal: Principal = Depends(get_user_principal),
    service: AttachmentService = Depends(get_attachment_service),
):
    result = service.download(attachment_id, principal)
    filename = result.attachment.original_filename
    suffix = filename.rsplit(".", 1)[-1] if "." in filename else "bin"
    disposition = f"attachment; filename=attachment.{suffix}; filename*=UTF-8''{quote(filename, safe='')}"
    return StreamingResponse(
        _stream_reader(result.reader),
        media_type=result.attachment.content_type,
        headers={
            "Content-Disposition": disposition,
            "Content-Length": str(result.attachment.size_bytes),
            "X-Content-Type-Options": "nosniff",
        },
    )


@router.delete("/api/v1/attachments/{attachment_id}", response_model=SuccessResponse[dict[str, bool]])
def delete_attachment(
    attachment_id: str,
    payload: AttachmentDelete,
    principal: Principal = Depends(get_user_principal),
    service: AttachmentService = Depends(get_attachment_service),
):
    service.delete(attachment_id, payload.reason, principal)
    return SuccessResponse(data={"deleted": True})

File: frontend/src/api/attachments.ts
import { api } from './client'
import type { ApiSuccess, AttachmentList, AttachmentType, AttachmentVisibility, TicketAttachment } from '../types'

export const attachmentKeys={ticket:(ticketId:string)=>['attachments',ticketId] as const}

export async function listAttachments(ticketId:string){
  return (await api.get<ApiSuccess<AttachmentList>>(`/tickets/${ticketId}/attachments`)).data.data
}

export async function uploadAttachment(ticketId:string,file:File,attachmentType:AttachmentType,visibility:AttachmentVisibility){
  const form=new FormData();form.append('file',file);form.append('attachment_type',attachmentType);form.append('visibility',visibility)
  return (await api.post<ApiSuccess<TicketAttachment>>(`/tickets/${ticketId}/attachments`,form,{timeout:60_000})).data.data
}

export async function deleteAttachment(attachmentId:string,reason:string){
  return (await api.delete<ApiSuccess<{deleted:boolean}>>(`/attachments/${attachmentId}`,{data:{reason}})).data.data
}

function responseFilename(disposition:string|undefined,fallback:string){
  const encoded=disposition?.match(/filename\*=UTF-8''([^;]+)/i)?.[1]
  if(!encoded)return fallback
  try{return decodeURIComponent(encoded)}catch{return fallback}
}

export async function downloadAttachment(attachment:TicketAttachment){
  const response=await api.get<Blob>(`/attachments/${attachment.id}/download`,{responseType:'blob',timeout:60_000})
  const url=URL.createObjectURL(response.data)
  const anchor=document.createElement('a');anchor.href=url;anchor.download=responseFilename(response.headers['content-disposition'],attachment.original_filename)
  document.body.appendChild(anchor);anchor.click();anchor.remove();URL.revokeObjectURL(url)
}
